### PR TITLE
Merger and split journey should handle when we have no suppliers.

### PR DESCRIPTION
**Pull request by martinkimmings**

### CODE REVIEW

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs [hunk 1]
```diff
@@ -124,5 +102,46 @@ public async Task SetSupplierSection(Order order, Supplier supplier, Contact con
...
+                        && ci.PublishedStatus == PublicationStatus.Published
+                        && ci.Solution.FrameworkSolutions.Select(f => f.Framework).Distinct().Any(f => !f.IsExpired));
+
+            if (practiceReorganisationType != PracticeReorganisationTypeEnum.None)
+            {
+                query = query.Where(ci => ci.SupplierServiceAssociations.Any(ssa => (ssa.AssociatedService.PracticeReorganisationType & practiceReorganisationType) == practiceReorganisationType));
+            }
+            else
+            {
+                query = query.Where(ci => ci.SupplierServiceAssociations.Any());
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs:138]: `else` block can be removed if the `.Any` clause is added to the main `Where` clause on line 126. The result should be the same

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs [hunk 2]
```diff
@@ -124,5 +102,46 @@ public async Task SetSupplierSection(Order order, Supplier supplier, Contact con
 
             await dbContext.SaveChangesAsync();
         }
+
+        private IQueryable<Supplier> GetAllSuppliersFromBuyingCatalogue()
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs]: Based on the implementation, maybe something like `GetAllActiveSuppliers`

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs [hunk 3]
```diff
@@ -23,47 +23,26 @@ public sealed class SupplierService : ISupplierService
             this.orderService = orderService ?? throw new ArgumentNullException(nameof(orderService));
         }
 
-        public Task<List<Supplier>> GetAllSuppliersFromBuyingCatalogue()
+        public Task<List<Supplier>> GetAllSuppliersByOrderType(OrderType orderType)
```
Comment (ghost) [inline src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs]: What about `GetAllSuppliersSupportingOrderType`, though it's a bit wordy

**Review by bjss-colin — CHANGES_REQUESTED** [review]

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs (hunk 1 above)
Comment (martinkimmings) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs:138]: Unless I'm misunderstanding, doesn't this mean we'd be doing this for mergers and splits.

```
.Where(
  ci =>
  ci.CatalogueItemType == CatalogueItemType.Solution
  && ci.PublishedStatus == PublicationStatus.Published
  && ci.Solution.FrameworkSolutions.Select(f => f.Framework).Distinct().Any(f => !f.IsExpired)
  && ci.SupplierServiceAssociations.Any()
  && ci.SupplierServiceAssociations.Any(ssa => (ssa.AssociatedService.PracticeReorganisationType & practiceReorganisationType) == practiceReorganisationType));
```

Not suggesting it affects the results but it felt unnecessary to do both. Might even result in more complicated sql.

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs (hunk 3 above)
Comment (martinkimmings) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs]: Maybe `ByOrderType` is superfluous as it's implied by the parameter. How about `GetActiveSuppliers` and `HasActiveSuppliers`. This is inline with renaming the private methods to `GetActiveSuppliers...`

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs (hunk 2 above)
Comment (martinkimmings) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs]: I've gone with `GetActiveSuppliersForSolutions` and `GetActiveSuppliersForSolutionsWithAssociatedServices`

File: src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs (hunk 1 above)
Comment (ghost) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs:138]: I was thinking more along the lines of this for the base query

```csharp
.Where(
  ci =>
  ci.CatalogueItemType == CatalogueItemType.Solution
  && ci.PublishedStatus == PublicationStatus.Published
  && ci.Solution.FrameworkSolutions.Select(f => f.Framework).Distinct().Any(f => !f.IsExpired)
  && ci.SupplierServiceAssociations.Any())
```

Comment (martinkimmings) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs:138]: Yeah sorry I've not explained myself well.  I meant the base query you suggest with the additional where clause of merger and split condition would effectively result in this - that's my understanding.

```
.Where(
  ci =>
  ci.CatalogueItemType == CatalogueItemType.Solution
  && ci.PublishedStatus == PublicationStatus.Published
  && ci.Solution.FrameworkSolutions.Select(f => f.Framework).Distinct().Any(f => !f.IsExpired)
  && ci.SupplierServiceAssociations.Any()
  && ci.SupplierServiceAssociations.Any(ssa => (ssa.AssociatedService.PracticeReorganisationType & practiceReorganisationType) == practiceReorganisationType))
```

Comment (ghost) [reply src/NHSD.GPIT.BuyingCatalogue.Services/Orders/SupplierService.cs:138]: Fair point, that is true

**Review by bjss-colin — APPROVED** [review]

**Review by imranbashir01 — APPROVED** [review]
